dds/DCPS/PublisherImpl.cpp

```diff
@@ -328,11 +314,53 @@ PublisherImpl::lookup_datawriter(const char* topic_name)
   }
 }
 
-DDS::ReturnCode_t
-PublisherImpl::delete_contained_entities()
+bool PublisherImpl::prepare_to_delete_datawriters()
 {
-  // mark that the entity is being deleted
-  set_deleted(true);
+  ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex, guard, pi_lock_, false);
+  bool result = true;
+  const DataWriterMap::iterator end = datawriter_map_.end();
+  for (DataWriterMap::iterator i = datawriter_map_.begin(); i != end; ++i) {
+    DataWriterImpl* const writer = dynamic_cast<DataWriterImpl*>(i->second.in());
+    if (!writer) {
+      result = false;
+    }
+    writer->prepare_to_delete();
```

Comment:
```suggestion
    if (writer) {
      writer->prepare_to_delete();
    } else {
      result = false;
    }
```

Reply:
I'm not sure why I did this one differently from the same situation from the method below...